hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadIListP.java

```diff
@@ -52,18 +64,40 @@ public boolean isCooperative() {
     @Override
     protected void init(@Nonnull Context context) {
         HazelcastInstance instance;
+        SerializationService serializationService;
         if (isRemote()) {
             instance = client = newHazelcastClient(asClientConfig(clientXml));
+            serializationService = ((HazelcastClientProxy) instance).getSerializationService();
         } else {
             instance = context.jetInstance().getHazelcastInstance();
+            serializationService = ((ProcCtx) context).serializationService();
         }
-        IList<Object> list = instance.getList(name);
-        final int size = list.size();
-        traverser = size <= FETCH_SIZE ?
-                traverseIterable(list)
-                :
+        traverser = createTraverser(instance, name).map(serializationService::toObject);
+    }
+
+    @SuppressWarnings("rawtypes")
+    private Traverser<Data> createTraverser(HazelcastInstance instance, String name) {
+        IList<Data> list = instance.getList(name);
+        int size = list.size();
+
+        if (list instanceof ClientListProxy) {
+            ClientListProxy proxy = (ClientListProxy) list;
+            return createTraverser(size, proxy::dataIterator, proxy::dataSubList);
+        } else if (list instanceof ListProxyImpl) {
+            ListProxyImpl proxy = (ListProxyImpl) list;
+            return createTraverser(size, proxy::dataIterator, proxy::dataSubList);
+        } else {
+            throw new RuntimeException("Unexpected list class: " + list.getClass().getName());
+        }
+    }
+
+    private Traverser<Data> createTraverser(int size,
+                                            Supplier<Iterator<Data>> iteratorSupplier,
+                                            BiFunction<Integer, Integer, List<Data>> subListSupplier) {
+        return size <= FETCH_SIZE ?
```

Comment:
did we really need both data iterator and sublist here? seems unnecessary

Reply:
I wanted to keep the existing behavior but maybe it could be simplified indeed.... :thinking: